xtermset-0.5.2/doc/xtermset.1
.TH XTERMSET 1 "Dec 2000" "User Commands"
.SH NAME
xtermset \- change settings of an xterm
.SH SYNOPSIS
xtermset [\-option ...]
.SH DESCRIPTION
.B xtermset
allows you to change the characteristics of an xterm window from the
command line. Most options have the same names as those that you would
give xterm at startup.
.SH OPTIONS
.TP 8
.I \-store [<filename>]
This option will save all the other command line options given to the
filename. If filename is omitted then the options will be written to
~/.xtermsetrc.
.TP 8
.I \-default [<filename>]
This option will load options from the specified filename. If the filename
is omitted then the options will be loaded from ~/.xtermsetrc. Options
specified on the command line preceed those that are in the options file.
.TP 8
.I \-f, \-force
xtermset normally checks if you are using it on a correct terminal (xterm
(derivative) or dtterm (derivative). With this option you can override
this checking. 
.TP 8
.I \-T string, \-title string
Sets the window title.
.TP 8
.I \-n string
Sets the icon name.
.TP 8
.I \-fg color
Sets the VT100 foreground color.
.TP 8
.I \-bg color
Sets the VT100 background color.
.TP 8
.I \-mousefg color
Sets the mouse pointer foreground color.
.TP 8
.I \-mousebg color
Sets the mouse pointer background color.
.TP 8
.I \-tekfg color
Sets the Tektronix emulation window foreground color.
.TP 8
.I \-tekbg color
Sets the Tektronix emulation window background color.
.TP 8
.I \-cr color
Sets the text cursor color.
.TP 8
.I \-hc color
Sets the color for highlighted (selected) text.
.TP 8
.I \-fn font-spec, \-font font-spec
Sets the font.
.TP 8
.I \-geom geometry-spec, \-geometry geometry-spec
Sets the window size and position; see X(1) for details.
.SH FILES
~/.xtermsetrc
.SH SEE ALSO
bash(1), xterm(1), X(1)
.SH BUGS
.I \-geom
does not support offsets from the left or bottom of the screen.
If the
.I geometry-spec
isn't in the correct format, the output will be strange.
.SH AUTHORS
Breyten Ernsting (bje@dds.nl)
.br
Decklin Foster (decklin@home.com)


debian/changelog
xtermset (0.5.2-6) unstable; urgency=medium

  * New maintainer (Closes: #614803)
  * Update S-V to 3.9.6
  * Fix array bounds overruns. Thanks David Starner <prosfilaes@gmail.com>
    for the patch (Closes: #713037)
  * Lintian cleaning

 -- Chrysostomos Nanakos <cnanakos@debian.org>  Wed, 29 Jul 2015 14:20:32 +0300

xtermset (0.5.2-5) unstable; urgency=low

  * QA upload.
  * Remove deprecated dpatch and upgrade to packaging format "3.0 quilt"
    (Closes: #664321).
  * Update to Standards-Version to 3.9.3.1 and debhelper to 9.
  * Add missing build-arch, build-indep rules and use dh_prep.

 -- Jari Aalto <jari.aalto@cante.net>  Wed, 28 Mar 2012 22:14:13 +0300

xtermset (0.5.2-4) unstable; urgency=low

  * Orphan this package

 -- Patrick Schoenfeld <schoenfeld@debian.org>  Wed, 23 Feb 2011 15:32:32 +0100

xtermset (0.5.2-3) unstable; urgency=low

  * Updated maintainer email address 
  * Updated Standards-Version to 3.9.0:
     + Added README.source to document patch handling 
  * Fix a spelling mistake in the last changelog entry

 -- Patrick Schoenfeld <schoenfeld@debian.org>  Tue, 03 Mar 2009 10:59:30 +0100

xtermset (0.5.2-2) unstable; urgency=low

  * Run autoreconf without '-s' to avoid FTBFS errors when building
    twice in a row
  * Include a copyright information with the year to debian/copyright
  * Remove files generated by autoreconf in clean target to avoid different
    results when building twice in a row
  * Run autoreconf in an own target configure so that building does not fail,
    because of the missing configure script
  * Add dpatch patch system
  * Add a patch for the wrong use of hyphens in the manpage
  * Add missing copyright information to debian/copyright
  * Add a copyright and licensing information for the Debian packaging to
    debian/copyright
  * Install upstream changelog
  * Install upstream documents NEWS and THANKS

 -- Patrick Schoenfeld <schoenfeld@in-medias-res.com>  Tue, 26 Feb 2008 11:09:51 +0100

xtermset (0.5.2-1) unstable; urgency=low

  * New maintainer (Closes: #455009)
  * New upstream release (Closes: #262302)
  * Run autoreconf to fix missing autoconf files
  * Updated Standards-Version to 3.7.3
  * Added Vcs-Headers
  * Updated copyright file to include include license excerpt
  * Added a watch file

 -- Patrick Schoenfeld <schoenfeld@in-medias-res.com>  Fri, 14 Dec 2007 21:05:42 +0100

xtermset (0.5.1-1) unstable; urgency=low

  * New upstream release, new contact information. Closes: #143730.
  * Adapt packaging to upstream changes: autoconf, more READMEish files.
  * Remove local variables from debian/changelog.
  * Get rid of /usr/doc compatibility.
  * Upgraded standards version to 3.5.9.
  * Use dpkg-gencontrol -isp to get Priority into the deb control file.

 -- Tommi Virtanen <tv@debian.org>  Sun, 20 Apr 2003 19:33:43 +0300

xtermset (0.4-2) unstable; urgency=low

  * Lowercased Eterm in Depends. Closes: #54886.

 -- Tommi Virtanen <tv@debian.org>  Fri, 14 Jan 2000 18:46:48 +0200

xtermset (0.4-1) unstable; urgency=low

  * Initial Release.

 -- Tommi Virtanen <tv@debian.org>  Mon,  1 Nov 1999 20:06:46 +0200




xtermset-0.5.2/src/xtermset.c
/* xtermset: change some settings of an xterm from the command line */

/*
** Copyright (C) 2000 Breyten J. Ernsting <bje@dds.nl>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
**
*/

/* this was orginally 'setxtitle' in the splitvt package,
 * created by Sam Lantinga(slouken@cs.ucdavis.edu)
 * rewritten by Breyten Ernsting (bje@dds.nl)
 * and again by Decklin Foster (decklin@home.com) */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "xtermset.h"

int main(int argc, char *argv[])
{
    int i,style,code,switches[3];
    char fn[3][MAXFNCHARS-1],c,*tok;
    char line[MAXOPTCHARS+10]; /* ? */
    FILE *rcfile=NULL;
    argument *argptr = NULL;

    for(i=0;i<3;i++) {
       switches[i]=0;
       fn[i][0]='\0';
    }

    if(argc<=2) { /* one command line option, make gnu comliant */
       if(argc==1) {
	  version();
          usage();
          exit(1);
       } else {
	  if(strcmp(argv[1],"--help")==0) {
	     version();
             usage();
	     gplinfo();
             exit(0);
          } else if (strcmp(argv[1],"--version")==0) {
	     version();
             exit(0);
	  }
       }
    }

    term_loadall();

    for (i=1; i<argc; i++) {
	switch (style=nextopt(argv[i], &code)) {
	    case BasicOption:
                MALLOCOPTION
		CHECKFORCOMMAND
		break;
	    case GeomOption:
                MALLOCOPTION
		CHECKFORCOMMAND
		break;
            case ShortOption:
                MALLOCOPTION
                break;
	    case SwitchOption:
                MALLOCOPTION
	        switches[code]=1;
                break;
            case ComplexSwitchOption:
                switches[code]=1;
                if ((i + 1 < argc) && (strncmp(argv[i+1],"-",1)!=0)) {
                    strcpy(fn[code],argv[++i]);
		}
                break;
	    case UnknownOption:
	        version();
		usage();
                argument_freeall();
                term_freeall();
		exit(2);
	}
    }

    if(switches[cswDefault]) {
        rcfile = open_rc_file(fn[cswDefault],"r");
        if(rcfile) {
           while(!feof(rcfile)) {
              line[0]='\0';
              for(i=0;!feof(rcfile);i++) {
                 c=getc(rcfile);
                 if ((c==EOF)||(c=='\n')) break;
                 line[i]=c;
              }
              line[i]='\0';
              if((line[0]=='#')||(strlen(line)==0))
                 continue;
              tok=strtok(line," ");
              if(tok==NULL) continue;
              style=atoi(tok);
              tok=strtok(NULL," ");
              if(tok==NULL) continue;
              code=atoi(tok);

	      argument_first();
	      argptr = argument_find(style,code);
              if(argptr==NULL) {
                 MALLOCOPTION
                 if((tok=strtok(NULL,"\n"))!=NULL)
                    strcpy(argptr->param,tok);
              }

              switch(argptr->style) {
                 case SwitchOption:
                    switches[code]=1;
                    break;
              }
           }
           fclose(rcfile);
        }
    }

    if(switches[cswStore])
        rcfile = open_rc_file(fn[cswStore],"w");

    if(!switches[0]) switches[0] = is_good_term();

    argptr = argument_first();
    while(argptr) {
       if(switches[cswStore] && (rcfile!=NULL)) {
           fprintf(rcfile,"%d %d",current->style,current->command);
           if(strlen(argptr->param)>0)
               fprintf(rcfile," %s",argptr->param);
           fprintf(rcfile,"\n");
       }

       switch (argptr->style) {
	   case BasicOption:
	       if(strlen(argptr->param)>0) {
                   CHECK {
                       printf(ESC "]%d;%s" BEL, argptr->command,argptr->param);
		   }
	       }
	       break;
           case GeomOption:
               if(strlen(argptr->param)>0) {
	           CHECK {
		       set_geom(argptr->param);
		   }
               }
	       break;
	case ShortOption:
	       CHECK {
		   printf(ESC "[%dt", argptr->command);
               }
	       break;
       }
       argptr = argument_next();
    }

    if(switches[cswStore])
        fclose(rcfile);

    term_freeall();
    argument_freeall();

    exit(0);
}

void version() {
    printf("GPF " PACKAGE " " VERSION "\n");
}

void usage()
{
    printf("\n  [-store [<filename>]] [-default [<filename>]]\n");
    printf("  [-force|f] [-title|T window-title] [-n icon-name]\n");
    printf("  [-fg color] [-bg color] [-cr color]\n");
    printf("  [-mousefg color] [-mousebg color]\n");
    printf("  [-tekfg color] [-tekbg color] [-hc color]\n");
    printf("  [-fn|font font-spec] [-geom size-spec]\n");
    printf("  [-iconify] [-restore] [-refresh]\n\n");
    printf("Report bugs to bje@dds.nl.\n");
}

void gplinfo() {
  printf("\nCopyright (C) 2000 Breyten J. Ernsting <bje@dds.nl>\n\n");
  printf("This program is free software; you can redistribute it and/or modify\n");
  printf("it under the terms of the GNU General Public License as published by\n");
  printf("the Free Software Foundation; either version 2 of the License, or\n");
  printf("(at your option) any later version.\n\n");
  printf("This program is distributed in the hope that it will be useful,\n");
  printf("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
  printf("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
  printf("GNU General Public License for more details.\n\n");
  printf("You should have received a copy of the GNU General Public License\n");
  printf("along with this program; if not, write to the Free Software\n");
  printf("Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.\n\n");
}

int nextopt(char *arg, int *code_return)
{
    OPTION("title", boTitle, BasicOption)
    OPTION("T", boTitle, BasicOption)
    OPTION("n", boIconName, BasicOption)
    OPTION("fg", boForeGround, BasicOption)
    OPTION("bg", boBackGround, BasicOption)
    OPTION("cr", boTextCursor, BasicOption)
    OPTION("mousefg", boMouseForeGround, BasicOption)
    OPTION("mousebg", boMouseBackGround, BasicOption)
    OPTION("tekfg", boTEKForeGround, BasicOption)
    OPTION("tekbg", boTEKBackGround, BasicOption)
    OPTION("hc", boHighlightCursor, BasicOption)
    OPTION("font", boFont, BasicOption)
    OPTION("fn", boFont, BasicOption)
    OPTION("geom", goGeom, GeomOption)
    OPTION("geometry", goGeom, GeomOption)
    OPTION("iconify", soIconify, ShortOption)
    OPTION("restore", soRestore, ShortOption)
    OPTION("refresh", soRefresh, ShortOption)
    OPTION("force",swForce, SwitchOption)
    OPTION("f",swForce,SwitchOption)
    OPTION("store",cswStore,ComplexSwitchOption)
    OPTION("default",cswDefault,ComplexSwitchOption)
    return UnknownOption;
}

void set_geom(char *geometry)
{
    char *x, *y, *width, *height;

    if (strncmp(geometry, "+", 1) == 0) {
	x = strtok(geometry, "+");
	y = strtok(NULL, "");
	printf(ESC "[3;%s;%st", x, y);
    } else {
	width = strtok(geometry, "x");
	height = strtok(NULL, "+");
	printf(ESC "[8;%s;%st", height, width);
	if ((x = strtok(NULL, "+"))) {
	    y = strtok(NULL, "");
	    printf(ESC "[3;%s;%st", x, y);
	}
    }
}

FILE *open_rc_file(char *fn,char *mode)
{
    char fname[MAXFNCHARS-1];
    if(strlen(fn)>0) {
        strcpy(fname,fn);
    } else {
        strcpy(fname,strcat(strcat(getenv("HOME"),"/"),RCFILE));
    }
    return fopen(fname,mode);
}
